pkg/capi/aws.go

```diff
@@ -0,0 +1,254 @@
+
+	providerSpec := &mapiv1.AWSMachineProviderConfig{}
+	Expect(yaml.Unmarshal(machineSet.Spec.Template.Spec.ProviderSpec.Value.Raw, providerSpec)).To(Succeed(), "it should be able to unmarshal the raw yaml into providerSpec")
+
+	return machineSet, providerSpec
+}
+
+func newAWSMachineTemplate(mapiProviderSpec *mapiv1.AWSMachineProviderConfig) *awsv1.AWSMachineTemplate {
+	By("Creating AWS machine template")
+
+	Expect(mapiProviderSpec).ToNot(BeNil(), "expected the mapi ProviderSpec to not be nil")
+	Expect(mapiProviderSpec.IAMInstanceProfile).ToNot(BeNil(), "expected the mapi IAMInstanceProfile to not be nil")
+	Expect(mapiProviderSpec.IAMInstanceProfile.ID).ToNot(BeNil(), "expected the mapi IAMInstanceProfile.ID to not be nil")
+	Expect(mapiProviderSpec.InstanceType).ToNot(BeEmpty(), "expected the mapi InstanceType to not be empty")
+	Expect(mapiProviderSpec.Placement.AvailabilityZone).ToNot(BeEmpty(), "expected the mapi Placement.AvailabilityZone to not be empty")
+	Expect(mapiProviderSpec.AMI.ID).ToNot(BeNil(), "expected the mapi AMI.ID to not be nil")
+	Expect(mapiProviderSpec.Subnet.Filters).ToNot(HaveLen(0), "expected the mapi Subnet.Filters to be present")
+	Expect(mapiProviderSpec.Subnet.Filters[0].Values).ToNot(HaveLen(0), "expected the mapi Subnet.Filters[0].Values to be present")
+	Expect(mapiProviderSpec.SecurityGroups).ToNot(HaveLen(0), "expected the mapi SecurityGroups to be present")
+	Expect(mapiProviderSpec.SecurityGroups[0].Filters).ToNot(HaveLen(0), "expected the mapi SecurityGroups[0].Filters to be present")
+	Expect(mapiProviderSpec.SecurityGroups[0].Filters[0].Values).ToNot(HaveLen(0), "expected the mapi SecurityGroups[0].Filters[0].Values to be present")
+
+	uncompressedUserData := true
+
+	awsMachineSpec := awsv1.AWSMachineSpec{
+		UncompressedUserData: &uncompressedUserData,
+		IAMInstanceProfile:   *mapiProviderSpec.IAMInstanceProfile.ID,
+		InstanceType:         mapiProviderSpec.InstanceType,
+		AMI: awsv1.AMIReference{
+			ID: mapiProviderSpec.AMI.ID,
+		},
+		Ignition: &awsv1.Ignition{
+			Version:     "3.4",
+			StorageType: awsv1.IgnitionStorageTypeOptionUnencryptedUserData,
+		},
+		Subnet: &awsv1.AWSResourceReference{
+			Filters: []awsv1.Filter{
+				{
+					Name:   "tag:Name",
+					Values: mapiProviderSpec.Subnet.Filters[0].Values,
+				},
+			},
+		},
+		AdditionalSecurityGroups: []awsv1.AWSResourceReference{
+			{
+				Filters: []awsv1.Filter{
+					{
+						Name:   "tag:Name",
+						Values: mapiProviderSpec.SecurityGroups[0].Filters[0].Values,
+					},
+				},
+			},
+		},
+	}
+
+	awsMachineTemplate := &awsv1.AWSMachineTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      awsMachineTemplateName,
+			Namespace: framework.ClusterAPINamespace,
+		},
+		Spec: awsv1.AWSMachineTemplateSpec{
+			Template: awsv1.AWSMachineTemplateResource{
+				Spec: awsMachineSpec,
+			},
+		},
```

Comment:
here can use resourcebuilder https://github.com/openshift/cluster-api-actuator-pkg/tree/master/testutils/resourcebuilder

Reply:
Updated, thank you @sunzhaohua2 PTAL again, it passed in local.
```
liuhuali@Lius-MacBook-Pro cluster-api-actuator-pkg % ./hack/ci-integration.sh -focus "Cluster API AWS MachineSet" -v
Running Suite: Machine Suite - /Users/liuhuali/project/cluster-api-actuator-pkg/pkg
===================================================================================
Random Seed: 1732506965

Will run 2 of 44 specs
------------------------------
[BeforeSuite] 
/Users/liuhuali/project/cluster-api-actuator-pkg/pkg/e2e_test.go:63
[BeforeSuite] PASSED [1.360 seconds]
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
Cluster API AWS MachineSet should be able to run a machine with a default provider spec [capi, disruptive]
/Users/liuhuali/project/cluster-api-actuator-pkg/pkg/capi/aws.go:100
  STEP: Creating core cluster @ 11/25/24 11:56:29.931
  STEP: Creating AWS machine template @ 11/25/24 11:56:30.766
  STEP: Creating MachineSet "aws-machineset-51071" @ 11/25/24 11:56:31.32
  STEP: Waiting for MachineSet machines "aws-machineset-51071" to enter Running phase @ 11/25/24 11:56:31.602
  STEP: Deleting MachineSet "aws-machineset-51071" @ 11/25/24 12:00:17.049
  STEP: Waiting for MachineSet "aws-machineset-51071" to be deleted @ 11/25/24 12:00:17.327
  STEP: Deleting /aws-machine-template @ 11/25/24 12:01:21.212
• [294.166 seconds]
------------------------------
Cluster API AWS MachineSet should be able to run a machine with cluster placement group [capi, disruptive]
/Users/liuhuali/project/cluster-api-actuator-pkg/pkg/capi/aws.go:112
I1125 12:01:24.951211   28912 aws_client.go:72] The created placementGroupID is pg-0b86fc1ccb6aa9742
  STEP: Creating AWS machine template @ 11/25/24 12:01:24.951
  STEP: Creating MachineSet "aws-machineset-75395" @ 11/25/24 12:01:25.238
  STEP: Waiting for MachineSet machines "aws-machineset-75395" to enter Running phase @ 11/25/24 12:01:25.521
  STEP: Deleting MachineSet "aws-machineset-75395" @ 11/25/24 12:05:51.76
  STEP: Waiting for MachineSet "aws-machineset-75395" to be deleted @ 11/25/24 12:05:52.055
  STEP: Deleting /aws-machine-template @ 11/25/24 12:06:55.93
• [336.224 seconds]
------------------------------
SSSSSSSSSSSSSSSS
------------------------------
[ReportAfterSuite] Autogenerated ReportAfterSuite for --junit-report
autogenerated by Ginkgo
[ReportAfterSuite] PASSED [0.003 seconds]
------------------------------

Ran 2 of 44 Specs in 631.752 seconds
SUCCESS! -- 2 Passed | 0 Failed | 0 Pending | 42 Skipped
PASS

Ginkgo ran 1 suite in 10m52.059989621s
Test Suite Passed

```